File: examples/statusline/statusline.mjs
import { basename } from "path";
import { execSync } from "child_process";

// Color helpers
const C = {
  reset: "\x1b[0m",
  dim: "\x1b[2m",
  bold: "\x1b[1m",
  cyan: "\x1b[36m",
  green: "\x1b[32m",
  yellow: "\x1b[33m",
  orange: "\x1b[38;5;208m",
  red: "\x1b[31m",
  redBlink: "\x1b[5;31m",
};

function colorPct(val, thresholds = [40, 60, 85]) {
  if (val == null) return "?";
  const n = Math.round(val);
  const [low, warn, crit] = thresholds;
  if (n >= crit) return `${C.red}${n}%${C.reset}`;
  if (n >= warn) return `${C.yellow}${n}%${C.reset}`;
  if (n >= low) return `${C.green}${n}%${C.reset}`;
  return `${n}%`;
}

function contextBar(usedPct) {
  const used = Math.max(0, Math.min(100, Math.round(usedPct)));
  const filled = Math.floor(used / 10);
  const bar = "\u2588".repeat(filled) + "\u2591".repeat(10 - filled);
  let numColor;
  if (used < 50) numColor = C.green;
  else if (used < 65) numColor = C.yellow;
  else if (used < 80) numColor = C.orange;
  else numColor = C.red;
  return `${C.dim}${bar}${C.reset} ${numColor}${used}%${C.reset}`;
}

let input = "";
process.stdin.on("data", (chunk) => (input += chunk));
process.stdin.on("end", () => {
  try {
    const d = JSON.parse(input);

    const cwd = d.cwd || d.workspace?.current_dir || ".";
    const dir = basename(cwd);
    let branch = d.worktree?.branch || "";
    if (!branch) {
      try {
        branch = execSync("git rev-parse --abbrev-ref HEAD", {
          cwd,
          encoding: "utf8",
          stdio: ["pipe", "pipe", "pipe"],
        }).trim();
      } catch {}
    }
    const model = d.model?.display_name || d.model?.id || "?";
    const ctx = d.context_window?.used_percentage;
    const fiveH = d.rate_limits?.five_hour?.used_percentage;
    const week = d.rate_limits?.seven_day?.used_percentage;
    const vim = d.vim?.mode;

    const parts = [];
    parts.push(dir);
    if (branch) parts.push(`${C.cyan}${branch}${C.reset}`);
    parts.push(`\x1b[38;5;242m${model}${C.reset}`);
    if (ctx != null) parts.push(contextBar(ctx));
    parts.push(`5h:${colorPct(fiveH)}`);
    parts.push(`7d:${colorPct(week)}`);
    if (vim) parts.push(`[${vim}]`);

    process.stdout.write(parts.join("  "));
  } catch {
    process.stdout.write("statusline error");
  }
});

File: examples/statusline/statusline.py
#!/usr/bin/env python3
"""Claude Code statusline: dir, branch, model, context bar, 5h/7d quotas."""
from __future__ import annotations
import json
import subprocess
import sys
from pathlib import Path

# ANSI color helpers
RESET = "\033[0m"
DIM = "\033[2m"
CYAN = "\033[36m"
GREEN = "\033[32m"
YELLOW = "\033[33m"
ORANGE = "\033[38;5;208m"
RED = "\033[31m"
RED_BLINK = "\033[5;31m"


def color_pct(val: float | None, thresholds: tuple[int, int, int] = (40, 60, 85)) -> str:
    if val is None:
        return "?"
    n = round(val)
    low, warn, crit = thresholds
    if n >= crit:
        return f"{RED}{n}%{RESET}"
    if n >= warn:
        return f"{YELLOW}{n}%{RESET}"
    if n >= low:
        return f"{GREEN}{n}%{RESET}"
    return f"{n}%"


def context_bar(used_pct: float) -> str:
    used = max(0, min(100, round(used_pct)))
    filled = used // 10
    bar = "\u2588" * filled + "\u2591" * (10 - filled)
    if used < 50:
        num_color = GREEN
    elif used < 65:
        num_color = YELLOW
    elif used < 80:
        num_color = ORANGE
    else:
        num_color = RED
    return f"{DIM}{bar}{RESET} {num_color}{used}%{RESET}"


def git_branch(cwd: str) -> str | None:
    try:
        result = subprocess.run(
            ["git", "-C", cwd, "--no-optional-locks", "symbolic-ref", "--short", "HEAD"],
            capture_output=True, text=True, timeout=3,
        )
        return result.stdout.strip() or None
    except (subprocess.TimeoutExpired, FileNotFoundError):
        return None


def main() -> None:
    try:
        data = json.loads(sys.stdin.read())
    except (json.JSONDecodeError, ValueError):
        data = {}

    cwd = data.get("workspace", {}).get("current_dir") or data.get("cwd", "")
    model = (data.get("model") or {}).get("display_name") or (data.get("model") or {}).get("id", "?")
    ctx_used = (data.get("context_window") or {}).get("used_percentage")
    rate_limits = data.get("rate_limits") or {}
    five_h = (rate_limits.get("five_hour") or {}).get("used_percentage")
    weekly = (rate_limits.get("seven_day") or {}).get("used_percentage")
    vim_mode = (data.get("vim") or {}).get("mode")

    dirname = Path(cwd).name if cwd else "."
    branch = git_branch(cwd) if cwd else None

    parts = [dirname]
    if branch:
        parts.append(f"{CYAN}{branch}{RESET}")
    parts.append(f"\033[38;5;242m{model}{RESET}")
    if ctx_used is not None:
        try:
            parts.append(context_bar(float(ctx_used)))
        except (TypeError, ValueError):
            pass
    parts.append(f"5h:{color_pct(five_h)}")
    parts.append(f"7d:{color_pct(weekly)}")
    if vim_mode:
        parts.append(f"[{vim_mode}]")

    print("  ".join(parts), end="")


if __name__ == "__main__":
    main()
